### PR TITLE
Add EdgeColoring

### LINKED ISSUE #54: Edge coloring
If every vertex has degree <= d, a d+1-edge coloring can be found in O(NM) time using https://en.wikipedia.org/wiki/Misra_%26_Gries_edge_coloring_algorithm.

https://github.com/ludopulles/tcr/blob/master/code/graphs/MisraGries.cpp

### PR DESCRIPTION
**Pull request by simonlindholm**
Closes #54.

### CODE REVIEW

**Comment by Tyilo** [comment]
I just tested this on [`gamescheduling`](https://open.kattis.com/problems/gamescheduling) where it works.

File: stress-tests/graph/EdgeColoring.cpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+		int col = cols[i];
+		assert(cols[i] <= maxdeg);
+		for (int x : {ed[i].first, ed[i].second}) {
+			assert(!usedCols[x][col]);
+			usedCols[x][col] = 1;
+		}
+	}
+}
+
+vector<pii> randomGraph(int n, int m, mt19937& rng) {
```
Comment (Chillee) [inline stress-tests/graph/EdgeColoring.cpp:23]: might be worth moving something like this into "utilities".

File: content/graph/EdgeColoring.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	vector<vi> adj(N, vi(ncols, -1));
+	for (pii e : eds) {
+		tie(u, v) = e;
+		fan[0] = v;
+		loc.assign(ncols, 0);
+		int at = u, end = u, d, c = free[u], ind = 0, i = 0;
+		while (d = free[v], !loc[d] && (v = adj[u][d]) != -1)
+			loc[d] = ++ind, cc[ind] = d, fan[ind] = v;
+		cc[loc[d]] = c;
+		for (int cd = d; at != -1; cd ^= c ^ d, at = adj[at][cd])
```
Comment (Chillee) [inline content/graph/EdgeColoring.h:29]: `cd ^= c ^ d` :thinking:

**Comment by Chillee** [comment]
Looks good to me, I think the API is easy - it's just a black box :0

**Pull request merged**